Data/HashMap/Base.hs

```diff
@@ -199,11 +199,15 @@ instance Foldable.Foldable (HashMap k) where
     foldr f = foldrWithKey (const f)
 
 #if __GLASGOW_HASKELL__ >= 711
+-- | '<>' = 'union'
 instance (Eq k, Hashable k) => Semigroup (HashMap k v) where
   (<>) = union
   {-# INLINE (<>) #-}
 #endif
 
+-- | 'mempty' = 'empty'
```

Comment:
`mempty` doesn't require documentation unless it's hard to see (e.g., `ZipList`). Speaking of `ZipList`, if you search [`Control.Applicative`](https://hackage.haskell.org/package/base-4.10.1.0/docs/Control-Applicative.html) for `ZipList`, you will see just how absolutely wretchedly horrible non-short Haddocks get for class instances. I therefore think we should probably stick to just `'mappend' = 'union'`. I can edit your commit and merge if you don't want to deal with this any more.

Reply:
Ew, yeah I didn't realize how bad that is. Also, couldn't the longer `mempty` docs and examples go on the `mempty` definition itself instead of the typeclass instance? Then it would show up under the expandable part of the instance docs and look just like any other function docs.

Reply:
I would also argue that putting the description to the right of the typeclass name is kinda bizarre, having the description below would probably make longer documentation much more readable, that's a haddock formatting question though.

Reply:
This one I really have to disagree with, because Haskell newbies need all the help they can get. What's obvious to you is not obvious to everyone. The fact that I can type either `mempty` and `empty` and they do exactly the same thing is not universally obvious.

Reply:
@m-renaud, instance methods can't have Haddock documentation, presumably because there's no obvious place to display it. @chris-martin, as best I can tell, the only decent places to write extensive documentation of instances are the introductory section at the very top of the module and in the Haddock documentation for the type itself. I don't have a definitely formed opinion about which of these is less bad.

Reply:
@chris-martin, you may be right that `mempty` is not obvious to all Haskell beginners. But the unfortunate bias in the `HashMap` `Semigroup` instance (that we use `union` rather than `unionWith (<>)`) means that people generally shouldn't use this instance anyway.

Reply:
@m-renaud Hmm.... that looks new! I wonder if we *can* document instance methods now!

Reply:
Changing `(<>)` from `union` to `unionWith (<>)` would require a very, very long process if it were ever undertaken. We'd have to start by having an extensive discussion on the libraries list. Then we'd have to deprecate the `Semigroup` and `Monoid` instances, then remove them. Then we'd have to wait a few more years before putting in the new ones. Then we'd have to cross our fingers and hope no one gets subtle bugs by using old code or reading old documentation. I kind of think we should do it, but it won't actually help anyone till 2025 or something.